src/cache/cache.rs

```diff
@@ -203,9 +203,31 @@ pub trait Storage: Send + Sync {
 
 /// Get a suitable `Storage` implementation from the environment.
 pub fn storage_from_environment() -> Box<Storage> {
+    let endpoint_env_res = env::var("SCCACHE_ENDPOINT");
+    let mut endpoint = String::from("s3.amazonaws.com");
+    match endpoint_env_res {
+        Ok(ep) => endpoint = ep,
+        Err(e) => warn!("Failed to interpret endpoint: {:?}", e),
+    }
+    let region_env_res = env::var("SCCACHE_REGION");
+    let mut region = String::from(""); // Empty region means 'us-east-1'.
+    match region_env_res {
+        Ok(r) => region = format!("-{}", r),
+        Err(e) => warn!("Failed to interpret region: {:?}", e),
+    }
+    if region == "-us-east-1".to_string() {
+        region = String::from("");
+    }
+
     if let Ok(bucket) = env::var("SCCACHE_BUCKET") {
-        trace!("Trying S3Cache({})", bucket);
-        match S3Cache::new(&bucket) {
+        if endpoint.ends_with(".amazonaws.com") {
```

Comment:
Instead of doing a string comparison here, maybe you should just combine this with the `SCCACHE_ENDPOINT` check, so something like:

```rust
let endpoint = match env::var("SCCACHE_ENDPOINT") {
  Ok(endpoint) => format!("{}/{}", endpoint, bucket),
  _ => format!("{}.s3{}.amazonaws.com", bucket, region),
}
```

That would mean that if someone explicitly set `SCCACHE_ENDPOINT=http://s3.amazonaws.com/` we wouldn't use the bucket vhost, but that seems fine.